leakcanary-android/src/main/java/com/squareup/leakcanary/internal/ForegroundService.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (C) 2018 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.leakcanary.internal;
+
+import android.app.IntentService;
+import android.app.Notification;
+import android.content.Intent;
+import android.os.IBinder;
+import android.support.annotation.Nullable;
+
+public abstract class ForegroundService extends IntentService {
+
+  private final int notificationContentTitleResId;
+
+  public ForegroundService(String name, int notificationContentTitleResId) {
+    super(name);
+    this.notificationContentTitleResId = notificationContentTitleResId;
+  }
+
+  @Override
+  public void onCreate() {
+    super.onCreate();
+    Notification.Builder builder = new Notification.Builder(this)
+        .setContentTitle(getString(notificationContentTitleResId));
+    Notification notification = LeakCanaryInternals.buildNotification(this, builder);
+    startForeground(LeakCanaryInternals.WORKING_NOTIFICATION_ID, notification);
+  }
+
+  @Override protected void onHandleIntent(@Nullable Intent intent) {
+    try {
+      onHandleIntentInForeground(intent);
+    } finally {
+      stopForeground(true);
```

Comment:
If more than one intent is queued, it will exit foreground early and crash.

Reply:
Probably can check the startID?

Reply:
Interesting. 

https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/app/IntentService.java

=> IntentService calls stopSelf() after every message, so I followed that model without thinking too much. But since we're calling startForeground() from onCreate(), should we simply call stopForeground() from on Destroy?

Reply:
onDestroy is the right spot. If the service crashes, the OS will clear the notification at the foreground ID on your behalf.